### ISSUE TITLE
suppress false-positives for section 4.1.3 Empty blocks: may be concise

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#15658]
I have read documentation: https://checkstyle.org/styleguides/google-java-style-20220203/javaguide.html#s4.1.3-braces-empty-blocks
https://checkstyle.org/google_style.html#a4.1.3
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

child of #15338

Remove false-positives for `{ }`, for this we need to modify google_checks.xml and use some suppressions. Adding new suppressions might introduce some false-negatives but that's okay, we will fix them in other issue, we're only concerned about removing false-positives as part of this issue.

I have used suppression provided by @romani at: https://github.com/checkstyle/checkstyle/issues/15338#issuecomment-2343605037 and modified a bit.

Example:

**OG xml  vs Modified xml:**

```
$ diff -u google_checks.xml google_checks_updated.xml 
--- google_checks.xml   2024-09-13 02:07:16.885630800 +0530
+++ google_checks_updated.xml   2024-09-13 02:00:49.320921600 +0530
@@ -78,11 +78,6 @@
     <module name="NoLineWrap">
       <property name="tokens" value="PACKAGE_DEF, IMPORT, STATIC_IMPORT"/>
     </module>
-    <module name="EmptyBlock">
-      <property name="option" value="TEXT"/>
-      <property name="tokens"
-               value="LITERAL_TRY, LITERAL_FINALLY, LITERAL_IF, LITERAL_ELSE, LITERAL_SWITCH"/>
-    </module>
     <module name="NeedBraces">
       <property name="tokens"
                value="LITERAL_DO, LITERAL_ELSE, LITERAL_FOR, LITERAL_IF, LITERAL_WHILE"/>
@@ -144,6 +139,16 @@
       <message key="ws.notPreceded"
               value="WhitespaceAround: ''{0}'' is not preceded with whitespace."/>
     </module>
+    <module name="SuppressionXpathSingleFilter">
+      <property name="checks" value="WhitespaceAround"/>
+      <property name="query" value="//*[self::LITERAL_IF or self::LITERAL_ELSE or self::STATIC_INIT or self::LITERAL_TRY or self::LITERAL_CATCH]/SLIST[count(./*)=1]
+                                 | //*[self::STATIC_INIT or self::LITERAL_TRY or self::LITERAL_IF]//*[self::RCURLY][parent::SLIST[count(./*)=1]]"/>
+    </module>
+    <module name="RegexpSinglelineJava">
+      <property name="format" value="\{[ ]+\}"/>
+      <property name="message" value="Empty blocks should have no spaces. Empty blocks
+      may only be represented as '{}' when not part of a multi-block statement (4.1.3)"/>
+    </module>
     <module name="OneStatementPerLine"/>
     <module name="MultipleVariableDeclarations"/>
     <module name="ArrayTypeStyle"/>
```

**Java file:**

```java
/** some. */
public class EmptyBlocks {

  static {} // ok

  static { } // violation

  void foo() {

    if (false) {} // ok
    if (false) { } // violation

    if (true) {
    }

    if (true) {
    } else { } // violation is for not having curly braces like: `{ }` but it should be for K & R style

    if (true) {
    } else {} // false-negative, should follow K & R style

    if (true) {
    } else {
    }
  }

  void foo2() throws Exception {
    try (MyResource r = new MyResource()) { } // violation
    try (MyResource r = new MyResource()) {}
    try (MyResource r = new MyResource()) {;} // violation is for not having curly braces like: `{ }` but it should be for K & R style
    try (MyResource r = new MyResource()) {} catch (Exception expected) {} // false-negative, should follow K & R style
  }

  /** some. */
  public class MyResource implements AutoCloseable {
    /** some. */
    @Override
    public void close() throws Exception {
      System.out.println("Closed MyResource");
    }
  }
}
```

**Output log for OG xml:**

```
$ java -jar checkstyle-10.18.1-all.jar -c google_checks.xml EmptyBlocks.java
Starting audit...
[WARN] C:\checkstyle testing\EmptyBlocks.java:4:10: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:4:11: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:10:16: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:10:16: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:10:17: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:11:16: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:13:15: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:16:15: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:17:12: Empty else block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:19:15: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:20:12: Empty else block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:20:12: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:20:13: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:22:15: Empty if block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:23:12: Empty else block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:28:43: Empty try block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:29:43: Empty try block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:29:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:29:44: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:31:43: Empty try block. [EmptyBlock]
[WARN] C:\checkstyle testing\EmptyBlocks.java:31:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:31:44: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:31:73: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:31:74: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

**Output log for modified xml:**

```
$ java -jar checkstyle-10.18.1-all.jar -c google_checks_updated.xml EmptyBlocks.java
Starting audit...
[WARN] C:\checkstyle testing\EmptyBlocks.java:6: Empty blocks should have no spaces. Empty blocks       may only be represented as {} when not part of a multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:11: Empty blocks should have no spaces. Empty blocks       may only be represented as {} when not part of a multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:17: Empty blocks should have no spaces. Empty blocks       may only be represented as {} when not part of a multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:28: Empty blocks should have no spaces. Empty blocks       may only be represented as {} when not part of a multi-block statement (4.1.3) [RegexpSinglelineJava]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:43: '{' at column 43 should have line break after. [LeftCurly]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:43: WhitespaceAround: '{' is not followed by whitespace. Empty blocks                may only be represented as {} when not part of a multi-block statement (4.1.3) [WhitespaceAround]
[WARN] C:\checkstyle testing\EmptyBlocks.java:30:45: WhitespaceAround: '}' is not preceded with whitespace. [WhitespaceAround]
Audit done.
```

All the cases mentioned at https://github.com/checkstyle/checkstyle/issues/15338#issuecomment-2336553623

> Looks like we just need to allow:
>
> - empty static block
> - empty simple IF
> - empty simple try

are covered, lemme know if there are more out there. 

Also, if you guys can suggest some better test cases then it would be really helpful :D

**Comment by romani** [comment]
this issue is duplicate of https://github.com/checkstyle/checkstyle/issues/15338

lets send PR for https://github.com/checkstyle/checkstyle/issues/15338 and based on test results see what extra issue we need to  merge PR for https://github.com/checkstyle/checkstyle/issues/15338 quicker and deal with false-negatives problems separately.
